### PR TITLE
Make plugin library for navigation2

### PR DESCRIPTION
**Pull request by vinnamkim**
Related to https://github.com/ros-planning/navigation2/issues/931, https://github.com/rst-tu-dortmund/teb_local_planner/issues/154

 - Remove teb_controller
 - Follow nav2_core::LocalPlanner interfaces to make plugin teb_local_planner::TebLocalPlannerROS
 - Update launch script and README.md to launch TurtleBot3 simulations with Navigation2 accompanied teb_local_planner

Signed-off-by: vinnamkim <vinnam.kim@gmail.com>

### CODE REVIEW

**Comment by SteveMacenski** [comment]
Question: 
why were the messages added to a new package next to teb in dashing? I don't have an opposition to it, just curious. If the reason was technical not being able to link, I added instructions to `ros2_docs` on it because I also found it challenging. 

https://github.com/ros2/ros2_documentation/pull/339

File: teb_local_planner/CMakeLists.txt [hunk 1]
```diff
@@ -70,25 +69,6 @@ set(EXTERNAL_LIBS ${Boost_LIBRARIES} ${SUITESPARSE_LIBRARIES} ${G2O_LIBRARIES})
 include_directories(
   include
   ${Boost_INCLUDE_DIRS}
-  ${costmap_converter_INCLUDE_DIRS}
```
Comment (SteveMacenski) [inline teb_local_planner/CMakeLists.txt:73]: Nice, thanks for this

File: teb_local_planner/launch/teb_params.yaml [hunk 2]
```diff
@@ -51,43 +51,43 @@ bt_navigator:
     use_sim_time: True
     bt_xml_filename: "bt_navigator.xml"
 
-dwb_controller:
+bt_navigator_rclcpp_node:
```
Comment (SteveMacenski) [inline teb_local_planner/launch/teb_params.yaml:54]: Shouldn't think `yaml` file only contain the controller_server code?

File: teb_local_planner/launch/teb_view.rviz [hunk 3]
```diff
@@ -5,7 +5,6 @@ Panels:
     Property Tree Widget:
       Expanded:
```
Comment (SteveMacenski) [inline teb_local_planner/launch/teb_view.rviz:6]: I'm not sure you meant to commit your rviz config?

File: teb_local_planner/package.xml [hunk 4]
```diff
@@ -46,4 +45,4 @@
   <export>
     <build_type>ament_cmake</build_type>
```
Comment (SteveMacenski) [inline teb_local_planner/package.xml:46]: Your missing the pluginlib export, see this for reference https://github.com/ros-planning/navigation2/blob/master/nav2_navfn_planner/package.xml#L37

**Review by SteveMacenski — CHANGES_REQUESTED** [review]

File: teb_local_planner/package.xml (hunk 4 above)
Comment (vinnamkim) [reply teb_local_planner/package.xml:46]: Fixed.

File: teb_local_planner/launch/teb_view.rviz (hunk 3 above)
Comment (vinnamkim) [reply teb_local_planner/launch/teb_view.rviz:6]: Sure, I wanted to add `/teb_poses` and `/teb_markers` for the default visualization setting.

File: teb_local_planner/launch/teb_params.yaml (hunk 2 above)
Comment (vinnamkim) [reply teb_local_planner/launch/teb_params.yaml:54]: Is it possible to overwrite only for `controller_server` part?

**Comment by vinnamkim** [comment]
> Question:
> why were the messages added to a new package next to teb in dashing? I don't have an opposition to it, just curious. If the reason was technical not being able to link, I added instructions to `ros2_docs` on it because I also found it challenging.
> 
> [ros2/ros2_documentation#339](https://github.com/ros2/ros2_documentation/pull/339)

This is because I just followed how other ROS2 packages did. If you want, I'll reconstruct TEB package to include the messages.

**Comment by SteveMacenski** [comment]
No need to go through the effort, I was just curious if there was a specific reason.

File: teb_local_planner/launch/teb_params.yaml (hunk 2 above)
Comment (SteveMacenski) [reply teb_local_planner/launch/teb_params.yaml:54]: Well in practice all the stuff in that one massive nav2_params yaml shouldn't be done like that, they should be separate yaml files for each package. But yes you can include multiple yaml files, now how it deals with conflicts is unknown to me

File: teb_local_planner/launch/teb_view.rviz (hunk 3 above)
Comment (SteveMacenski) [reply teb_local_planner/launch/teb_view.rviz:6]: Ok, there was just a bunch of changes. All good

File: teb_local_planner/launch/teb_params.yaml (hunk 2 above)
Comment (yathartha3) [reply teb_local_planner/launch/teb_params.yaml:54]: Probably, we should scope the local controller (dwb/teb) parameters, eg: 

```
controller_server:
  ros__paramaters:
    teb.param1 = 1
    teb.param2 = 7

**Comment by yathartha3** [comment]
Just wondering, change all ```.h``` to ```.hpp```?

**Comment by croesmann** [comment]
Big thanks for your efforts @vinnamkim and @SteveMacenski to bring the teb to ros2. 
I definitely prefer the new plugin approach to the previous one and the dependencies on dwb etc. removed with it.

Regarding the new msg package: this is something that can still be discussed. I already thought about splitting the packages in ros1 as it can be advantageous for usage in a multiple-machine setup as the number of dependencies (navigation stack, libg2o, costmap_converter etc.) is large. On the other hand, having multiple packages just for a local planner could also be too fine-grained (meta-package, core planner and msgs). According to the initial ros2 version of @vinnamkim, it seams that we do not need a meta package in ros2, but can still have two packages in a single git repo and a common cmake file? So up to now, I agree to proceed this way.

@yathartha3: I know that headers in ros2 core packages use `.hpp`.
However, personally, I like to use `.h` for headers even in c++ and `.hpp` for definitions of template functions/methods. But if it is best practice for all ros2 packages to follow the `.hpp` style, I would agree.

**Comment by yathartha3** [comment]
How are we doing here?

**Review by yathartha3 — CHANGES_REQUESTED** [review]
@vinnamkim I found some dwb references in the code, e.g. in, 

_planner_interface.h:_
 ```virtual bool isTrajectoryFeasible(dwb_critics::ObstacleFootprintCritic* costmap_model, blah, blah) = 0;```

_and teb_local_planner_ros.h:_
```std::shared_ptr<dwb_critics::ObstacleFootprintCritic> costmap_model_;```

There might be more elsewhere. I understand this is from the first port (when not using plugin), when we didn't have the plugin interfaces.
However, now, I believe this plugin should be free of dwb references.

**Comment by croesmann** [comment]
I am going to create a new ros2 installation next week. Probably, we can push a teb ros2 release soon.

**Review by SteveMacenski — APPROVED** [review]

**Pull request merged**